### ISSUE TITLE
Regression in type narrowing for type variable with union bound

### ISSUE THREAD

**Issue by adampauls** [microsoft/pyright#6475]
**Describe the bug**
Narrowing a `type[T]` using `issubclass` used to work (as of 1.1.335) where `T` was bound by a union, now it appears not to. 

**Code or Screenshots**

```python
from typing import TypeVar

T = TypeVar("T", bound=int | str)  # no errors below for `bound=object`


def bar(x: type[T]) -> type[str]:
    if issubclass(x, str):
        reveal_type(x)  # [1.1.335] information: Type of "x" is "type[str]*" [1.1.336]  information: Type of "x" is "type[T@bar]"
        return x
    return str
```

Full error is
```
x.py:11:16 - error: Argument of type "type[T@bar]" cannot be assigned to parameter "__key" of type "type[str]" in function "__setitem__"
    Type "type[str]* | type[int]*" cannot be assigned to type "type[str]"
      "type[int]*" is incompatible with "type[str]"
      Type "type[int]*" cannot be assigned to type "type[str]" (reportGeneralTypeIssues)

```

**VS Code extension or command-line**
Command line, 1.1.336.

**Comment by adampauls** [comment]
The issue exists for `isinstance` as well:
```python
from typing import TypeVar

T = TypeVar("T", bound=int | str)


def bar(x: T) -> str:
    if isinstance(x, str):
        reveal_type(x)  # [1.1.335] information: Type of "x" is "type[str]*" [1.1.336]  information: Type of "x" is "type[T@bar]"
        return x
    return "5"
```

**Comment by erictraut** [comment]
This is due to the fix for [this issue](https://github.com/microsoft/pyright/issues/6434). As I said in the resolution of that issue, this change has a tradeoff. If I fix the other bug, it breaks this case and vice versa.

I'll need to think more about whether there's some way to handle both cases, but I don't currently see one — at least not without adding full support for intersection types to the type system, and that's not happening any time soon. The problem here is that the proper narrowed type is _both_ `type[str]` and `type[T]`. In other words, it's an intersection between the two types. Without intersection support, a type checker needs to narrow the type to _either_ `type[str]` _or_ `type[T]`, and either of those will result in false positive errors in some cases.

**Comment by adampauls** [comment]
Makes sense. I don't know the internals of upright at all, but is it possible to narrow the type to `T1=TypeVar("T1", str)`, but remember that `T1=T`?

**Comment by adampauls** [comment]
I'll also add that the second snippet of code is quite worrying:

```python
if isinstance(x, str):
        reveal_type(x)  # [1.1.335] information: Type of "x" is "type[str]*" [1.1.336]  information: Type of "x" is "type[T@bar]"
```

It's very surprising that `x` isn't known to be a `str` immediately after an isinstance check, and even more surprising that this only happens if you have a union bound on the type variable -- unbound or non-union-bounded type variables narrow in the expected way.

**Comment by erictraut** [comment]
This is addressed in pyright 1.1.337, which I just published. It will also be included in a future release of pylance.
